hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -630,7 +650,28 @@ case class TableJoin(left: TableIR, right: TableIR, joinType: String, joinKey: I
     val rightKeyFieldIdx = rightRVDType.kFieldIdx
     val leftValueFieldIdx = leftRVDType.valueFieldIdx
     val rightValueFieldIdx = rightRVDType.valueFieldIdx
-    val newRowPType = PType.canonical(newRowType).setRequired(true).asInstanceOf[PStruct]
+
+    val (lkT, lvT, rvT) = joinType match {
+      case "inner" =>
+        val keyTypeFields = castFieldRequiredeness(leftRVDType.kType, true)
+        (keyTypeFields, noIndex(leftRVDType.valueType.fields), noIndex(rightRVDType.valueType.fields))
+      case "left" =>
+        val rValueTypeFields = castFieldRequiredeness(rightRVDType.valueType, false)
+        (noIndex(leftRVDType.kType.fields), noIndex(leftRVDType.valueType.fields), rValueTypeFields)
+      case "right" =>
+        val keyTypeFields = castFieldToPType(leftRVDType.kType, rightRVDType.kType)
```

Comment:
castFieldToPType has a confusing name. Probably clearer to do something like the following inline:

```
leftRVDType.key.zip(rightRVDType.kType.types)
```